### PR TITLE
Chatty project systems

### PR DESCRIPTION
**Pull request by jrieken**
Things aren't always smooth so project system tell more explicit when things go bad (more than logging). This PR adds events for these situations
* no DNX or no matching version was found
* DTH reports, like malformed project.json
* MSBuild fails to parse csproject-file
* MSBuild fails to process csproject-file, eg missing extra targets

In all cases a ProjectStatusMessage is send via the event bus.

### CODE REVIEW

File: src/OmniSharp/AspNet5/PackagesRestoreTool.cs [hunk 1]
```diff
@@ -32,6 +36,11 @@ public void Run(Project project)
         {
             Task.Factory.StartNew(() =>
             {
+                this._emitter.Emit(EventTypes.PackageRestoreStarted, new PackageRestoreMessage()
```
Comment (nosami) [inline src/OmniSharp/AspNet5/PackagesRestoreTool.cs]: this. ;)

File: src/OmniSharp/AspNet5/PackagesRestoreTool.cs [hunk 2]
```diff
@@ -80,6 +89,12 @@ public void Run(Project project)
                                 HostId = _context.HostId
                             });
                         }
+
+                        this._emitter.Emit(EventTypes.PackageRestoreFinished, new PackageRestoreMessage()
```
Comment (nosami) [inline src/OmniSharp/AspNet5/PackagesRestoreTool.cs]: and this :)

File: src/OmniSharp/MSBuild/MSBuildProjectSystem.cs [hunk 3]
```diff
@@ -163,7 +167,15 @@ private ProjectFileInfo CreateProject(string projectFilePath)
             }
             catch (Exception ex)
             {
-                _logger.WriteWarning(string.Format("Failed to process project file '{0}'.", projectFilePath), ex);
+                var message = new ProjectStatusMessage()
+                {
+                    LogLevel = "Error",
+                    FileName = projectFilePath,
+                    Text = string.Format("Failed to process project file '{0}'.", projectFilePath),
```
Comment (nosami) [inline src/OmniSharp/MSBuild/MSBuildProjectSystem.cs]: string interpolation?

Comment (jrieken) [inline src/OmniSharp/MSBuild/MSBuildProjectSystem.cs]: I just kept what we had. What do you prefer?

Comment (nosami) [inline src/OmniSharp/MSBuild/MSBuildProjectSystem.cs]: It's not important. Just that string interpolation looks a little nicer.

File: src/OmniSharp/AspNet5/AspNet5Paths.cs [hunk 4]
```diff
@@ -18,6 +17,7 @@ public class AspNet5Paths
         private readonly OmniSharpOptions _options;
         private readonly ILogger _logger;
         public string RuntimePath { get; private set; }
+        public string RuntimePathErrorMessage { get; private set; }
```
Comment (davidfowl) [inline src/OmniSharp/AspNet5/AspNet5Paths.cs]: This is kinda weird

File: src/OmniSharp/AspNet5/AspNet5ProjectSystem.cs [hunk 5]
```diff
@@ -101,7 +106,7 @@ public void Initalize()
 
                     if (m.MessageType == "ProjectInformation")
                     {
-                        var val = m.Payload.ToObject<ProjectMessage>();
+                        var val = m.Payload.ToObject<Microsoft.Framework.DesignTimeHost.Models.OutgoingMessages.ProjectMessage>();
```
Comment (davidfowl) [inline src/OmniSharp/AspNet5/AspNet5ProjectSystem.cs]: Why?

File: src/OmniSharp/MSBuild/MSBuildLogger.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+        public void Shutdown()
+        {
+            foreach (var action in _callOnShutdown)
+            {
+                action();
+            }
+        }
+
+        private void OnError(object sender, Microsoft.Build.Framework.BuildErrorEventArgs args)
```
Comment (davidfowl) [inline src/OmniSharp/MSBuild/MSBuildLogger.cs]: Why fully qualified type names?

File: src/OmniSharp/MSBuild/MSBuildProjectSystem.cs [hunk 7]
```diff
@@ -9,6 +9,7 @@
 using Microsoft.CodeAnalysis.MSBuild;
 using Microsoft.CodeAnalysis.Text;
 using Microsoft.Framework.Logging;
+using OmniSharp.Models;
```
Comment (davidfowl) [inline src/OmniSharp/MSBuild/MSBuildProjectSystem.cs]: IEventEmitter isn't a model. If it's in that interface we should move it to services.

File: src/OmniSharp/Models/ProjectStatusMessage.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace OmniSharp.Models
+{
+    public class ProjectStatusMessage
+    {
+        public string FileName { get; set; }
+        public string LogLevel { get; set; }
+        public string Text { get; set; }
+        public string ExceptionMessage { get; set; }
+        public IEnumerable<DiagnosticLocation> Diagnostics { get; set; }
```
Comment (davidfowl) [inline src/OmniSharp/Models/ProjectStatusMessage.cs]: What type is DiagnosticLocation?

File: src/OmniSharp/AspNet5/AspNet5ProjectSystem.cs (hunk 5 above)
Comment (jrieken) [inline src/OmniSharp/AspNet5/AspNet5ProjectSystem.cs]: ops - had at one point a ProjectMessage from a different package

File: src/OmniSharp/MSBuild/MSBuildLogger.cs (hunk 6 above)
Comment (jrieken) [inline src/OmniSharp/MSBuild/MSBuildLogger.cs]: There is MSFT.Framework.ILogger and MSBuild.Framework.ILogger.

File: src/OmniSharp/MSBuild/MSBuildProjectSystem.cs (hunk 7 above)
Comment (jrieken) [inline src/OmniSharp/MSBuild/MSBuildProjectSystem.cs]: IEventEmitter is in services but the events are in models

File: src/OmniSharp/Models/ProjectStatusMessage.cs (hunk 8 above)
Comment (jrieken) [inline src/OmniSharp/Models/ProjectStatusMessage.cs]: It's the type we send out for diagnostics: https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp/Api/Diagnostics/OmnisharpController.Diagnostics.cs#L46

Comment (davidfowl) [inline src/OmniSharp/Models/ProjectStatusMessage.cs]: Do we really want to rely on the roslyn type?

Comment (jrieken) [inline src/OmniSharp/Models/ProjectStatusMessage.cs]: That's not a roslyn type

**Comment by jrieken** [comment]
this isn't ready

**Pull request closed without merging**